src/cpp/session/modules/SessionRmdNotebook.R

```diff
@@ -205,165 +185,122 @@ assign(".rs.notebookVersion", envir = .rs.toolsEnv(), "1.0")
+                  
+                  bodyEl <- .rs.extractHTMLBodyElement(val)
+                  
+                  rendered <- htmltools::htmlPreserve(bodyEl)
+                  annotated <- .rs.rnb.htmlAnnotatedOutput(rendered, "htmlwidget", meta = jsonContents)
+                  output <- knitr::asis_output(annotated, meta = jsonContents)
+                  return(output)
+               }
+               
+               # TODO: shouldn't get here?
+               NULL
+            })
+            
+            list(
+               structure(list(src = NULL), class = "source"),
+               htmlOutput
+            )
          }
-      })
       
-      # insert into document
-      injection <- paste(unlist(htmlList), collapse = "\n")
+      )
       
-      html[[i]] <- paste(injection, sep = "\n", collapse = "\n")
-      chunkIdx <- chunkIdx + 1
    }
    
-   # Inject JSON dependency information into document
-   # TODO: Resolve duplicates
-   htmlInjection <- unlist(lapply(htmlDependencies, function(dep) {
-      
-      injection <- c()
-      
-      # Inject a CSS resource
-      if (!is.null(dep$stylesheet)) {
-         cssPath <- file.path(dep$src$file, dep$stylesheet)
-         contents <- .rs.readFile(cssPath, binary = TRUE)
-         encoded <- .rs.base64encode(contents)
-         htmlAttributes <- list(
-            "data-rnb-id" = sprintf("lib/%s-%s/%s", dep$name, dep$version, dep$stylesheet),
-            "href" = sprintf("data:text/css;charset=utf8;base64,%s", encoded),
-            "rel" = "stylesheet",
-            "type" = "text/css"
-         )
-         cssHtml <- sprintf("<link %s />", .rs.listToHtmlAttributes(htmlAttributes))
-         injection <- c(injection, cssHtml)
-      }
-      
-      # Inject a JS resource
-      if (!is.null(dep$script))
-      {
-         jsPath <- file.path(dep$src$file, dep$script)
-         contents <- .rs.readFile(jsPath, binary = TRUE)
-         encoded <- .rs.base64encode(contents)
-         htmlAttributes <- list(
-            "data-rnb-id" = sprintf("lib/%s-%s/%s", dep$name, dep$version, dep$script),
-            "src" = sprintf("data:application/x-javascript;base64,%s", encoded)
-         )
-         scriptHtml <- sprintf("<script %s></script>", .rs.listToHtmlAttributes(htmlAttributes))
-         injection <- c(injection, scriptHtml)
-      }
-      
-      # Inject other header contents
-      rendered <- htmltools:::renderTags(dep)
-      headContent <- as.character(rendered$head)
-      if (length(headContent) && any(nzchar(headContent)))
-         injection <- c(injection, headContent)
-      
-      paste(injection, collapse = "\n")
-   }))
+   exit <- function(...) {
+      knitr::knit_hooks$set(knitHooks)
+   }
    
-   bodyIdx <- tail(grep("^\\s*</head>\\s*$", html, perl = TRUE), n = 1)
-   html <- c(
-      html[1:(bodyIdx - 1)],
-      htmlInjection,
-      html[bodyIdx:length(html)]
-   )
+   suppressMessages(trace(
```

Comment:
In theory formats can set all the hooks they need. You can specify hooks in
knitr options and they are installed after the call render_markdown:

https://github.com/rstudio/rmarkdown/blob/master/R/render.R#L291-L298

Perhaps you are saying that you actually want the defaults as well. I might
be missing something, let's discuss on Monday.


On Sat, Apr 30, 2016 at 4:04 PM, Kevin Ushey <notifications@github.com>
wrote:

> In src/cpp/session/modules/SessionRmdNotebook.R
> <https://github.com/rstudio/rstudio/pull/691#discussion_r61674320>:
>
> >
> > -   bodyIdx <- tail(grep("^\\s*</head>\\s*$", html, perl = TRUE), n = 1)
> > -   html <- c(
> > -      html[1:(bodyIdx - 1)],
> > -      htmlInjection,
> > -      html[bodyIdx:length(html)]
> > -   )
> > +   suppressMessages(trace(
>
> The reason I went this route was because (if I recall correctly):
>
>    1. rmarkdown calls knitr::render_markdown(), which sets a bunch of
>    hooks (https://github.com/rstudio/rmarkdown/blob/master/R/render.R#L260
>    )
>    2. I want to set a bunch of hooks in the output format,
>    3. The hooks I set clobber those set by render_markdown().
>
> By using trace(), I can capture the state of the knitr hooks before
> entering knitr, then add my own hooks which just call the original hooks
> + add some extra stuff.
>
> Alternatively, perhaps the output format needs to also use the hooks that
> are set by knitr::render_markdown().
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/691/files/344ad98062a258bce8419895c7a94ba58dd48d01#r61674320>
>